model/util.go

```diff
@@ -17,114 +17,246 @@ package model
 import (
 	"bytes"
 	"encoding/json"
+	"errors"
 	"fmt"
 	"net/http"
+	"net/url"
 	"os"
 	"path/filepath"
+	"reflect"
+	"sync/atomic"
+	"time"
 
 	"sigs.k8s.io/yaml"
-
-	"strings"
-	"sync/atomic"
 )
 
 // +k8s:deepcopy-gen=false
 
-const prefix = "file:/"
+// TODO: Remove global variable
+var httpClient = http.Client{Timeout: time.Duration(1) * time.Second}
+
+type UnmarshalError struct {
+	err           error
+	parameterName string
+	primitiveType reflect.Kind
+	objectType    reflect.Kind
+}
+
+func (e *UnmarshalError) Error() (message string) {
+	if e.err == nil {
+		panic("unmarshalError fail")
+	}
+
+	var syntaxErr *json.SyntaxError
+	var unmarshalTypeErr *json.UnmarshalTypeError
 
-// TRUE used by bool fields that needs a boolean pointer
-var TRUE = true
+	if errors.As(e.err, &syntaxErr) {
+		message = fmt.Sprintf("%s has a syntax error \"%s\"", e.parameterName, syntaxErr.Error())
 
-// FALSE used by bool fields that needs a boolean pointer
-var FALSE = false
+	} else if errors.As(e.err, &unmarshalTypeErr) {
+		var primitiveTypeName string
+		var objectTypeName string
+
+		if e.primitiveType != 0 {
+			primitiveTypeName = e.primitiveType.String()
+		}
+		if e.objectType != 0 {
+			switch e.objectType {
+			case reflect.Struct:
+				objectTypeName = "object"
+			case reflect.Map:
+				objectTypeName = "object"
+			case reflect.Slice:
+				objectTypeName = "array"
+			default:
+				objectTypeName = e.objectType.String()
+			}
+		}
+
+		if unmarshalTypeErr.Struct == "" && unmarshalTypeErr.Field == "" {
+			message = fmt.Sprintf("%s must be an %s or %s", e.parameterName, primitiveTypeName, objectTypeName)
+		} else if unmarshalTypeErr.Struct != "" && unmarshalTypeErr.Field != "" {
+
+			switch unmarshalTypeErr.Type {
+			case reflect.TypeOf(InvokeKindSync):
+				primitiveTypeName = fmt.Sprintf("%s,%s", InvokeKindSync, InvokeKindAsync)
+			case reflect.TypeOf(ForEachModeTypeSequential):
```

Comment:
This will be more complicated when adding a new type, and synchronous modifications are also needed here.

Reply:
The last commit `029d335` I add an interface to deal with validations and error handling.
```
type Kinds interface {
	AllKinds() []string
	String() string
}
```
That way the validation and error handling can get the same values in the method `AllKinds`.